applications/TrilinosApplication/external_includes/trilinos_monotonicity_preserving_solver.h

```diff
@@ -0,0 +1,412 @@
+//    |  /           |
+//    ' /   __| _` | __|  _ \   __|
+//    . \  |   (   | |   (   |\__ `
+//   _|\_\_|  \__,_|\__|\___/ ____/
+//                   Multi-Physics
+//
+//  License:		 BSD License
+//					 Kratos default license: kratos/license.txt
+//
+//  Main authors:    Daniel Diez
+//
+
+#if !defined(KRATOS_TRILINOS_MONOTONICITY_PRESERVING_SOLVER_H_INCLUDED )
+#define  KRATOS_TRILINOS_MONOTONICITY_PRESERVING_SOLVER_H_INCLUDED
+
+// System includes
+
+// External includes
+
+// Project includes
+#include "includes/define.h"
+#include "factories/linear_solver_factory.h"
+#include "linear_solvers/linear_solver.h"
+#include "utilities/parallel_utilities.h"
+
+namespace Kratos
+{
+
+///@name Kratos Globals
+///@{
+
+///@}
+///@name Type Definitions
+///@{
+
+///@}
+///@name  Enum's
+///@{
+
+///@}
+///@name  Functions
+///@{
+
+///@}
+///@name Kratos Classes
+///@{
+
+/**
+ * @class TrilinosMonotonicityPreservingSolver
+ * @ingroup KratosCore
+ * @brief
+ * @details
+ * @author Daniel Diex
+ * @tparam TSparseSpaceType The sparse space definition
+ * @tparam TDenseSpaceType The dense space definition
+ * @tparam TReordererType The reorder considered
+ */
+template<class TSparseSpaceType, class TDenseSpaceType,
+         class TReordererType = Reorderer<TSparseSpaceType, TDenseSpaceType> >
+class TrilinosMonotonicityPreservingSolver
+    : public LinearSolver<TSparseSpaceType, TDenseSpaceType,  TReordererType>
```

Comment:
```suggestion
    : public MonotonicityPreservingSolver<TSparseSpaceType, TDenseSpaceType,  TReordererType>
```

from what I see only `ProvideAdditionalData` is different, hence I think it should be possible to derive in this way

Reply:
Im afraid this does not work because then it tries to instantiate MonotonicityPreservingSolver with the trilinos space so you get errors like this:
```

/home/daniel/KratosDEV/KratosGitHub/applications/TrilinosApplication/custom_factories/trilinos_linear_solver_factory.cpp:105:1:   required from here
/home/daniel/KratosDEV/KratosGitHub/kratos/linear_solvers/monotonicity_preserving_solver.h:201:36: error: ‘Kratos::MonotonicityPreservingSolver<Kratos::TrilinosSpace<Epetra_FECrsMatrix, Epetra_FEVector>, Kratos::UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> >, Kratos::Reorderer<Kratos::TrilinosSpace<Epetra_FECrsMatrix, Epetra_FEVector>, Kratos::UblasSpace<double, boost::numeric::ublas::matrix<double>, boost::numeric::ublas::vector<double> > > >::SparseMatrixType {aka class Epetra_FECrsMatrix}’ has no member named ‘value_data’
         double *values_vector = rA.value_data().begin();
```

Reply:
you are right, I hadn't considered that...
Then I guess it is ok as is, just add the test then I can approve :)